src/OpenTelemetry.Api/Metrics/Timer.cs

```diff
@@ -0,0 +1,176 @@
+            : base(meter, name, "ms", description)
+        {
+        }
+
+        /// <summary>
+        /// Start a timer for a Timer instrument.
+        /// </summary>
+        /// <param name="tags">Attributes.</param>
+        /// <returns>Timer (<see cref="TimeMarkerWithTag{T}"/>) state.</returns>
+        public TimeMarkerWithTag<T> Start(params KeyValuePair<string, object>[] tags)
+        {
+            return new TimeMarkerWithTag<T>(this, tags);
+        }
+
+        /// <summary>
+        /// Stop and record time for a Timer instrument.
+        /// </summary>
+        /// <param name="mark">Timer state.</param>
+        public void Stop(TimeMarkerWithTag<T> mark)
+        {
+            if (mark.TimerMark.Timer == this)
+            {
+                mark.Dispose();
+            }
+            else
+            {
+                throw new Exception("Mismatched Timer!");
+            }
+        }
+
+        /// <summary>
+        /// Start a timer for a Timer instrument.
+        /// </summary>
+        /// <returns>Timer (<see cref="TimeMarker{T}"/>) state.</returns>
+        public TimeMarker<T> Start()
+        {
+            return new TimeMarker<T>(this);
+        }
+
+        /// <summary>
+        /// Stop and record time for a Timer instrument.
+        /// </summary>
+        /// <param name="mark">Timer state.</param>
+        /// <param name="tags">Attributes.</param>
+        public void Stop(TimeMarker<T> mark, params KeyValuePair<string, object>[] tags)
+        {
+            if (mark.Timer == this)
+            {
+                this.Record(mark.ElapsedMilliseconds, tags);
+            }
+            else
+            {
+                throw new Exception("Mismatched Timer!");
+            }
+        }
+
+        internal void Record(int elapsed, KeyValuePair<string, object>[] tags)
+        {
+            T value;
+
+            if (typeof(T) == typeof(int))
+            {
+                value = (T)(object)(int)elapsed;
+            }
+            else if (typeof(T) == typeof(long))
+            {
+                value = (T)(object)(long)elapsed;
+            }
+            else if (typeof(T) == typeof(double))
+            {
+                value = (T)(object)(double)elapsed;
+            }
+            else
+            {
+                throw new Exception("Unsupported Type");
+            }
+
+            this.RecordMeasurement(value, tags);
+        }
+
+        /// <summary>
+        /// TimeMarker records the start state of a Timer instrument.
+        /// </summary>
+        /// <typeparam name="T1">Support <c>int</c>, <c>long</c>, <c>double</c>.</typeparam>
+        public struct TimeMarker<T1>
+            where T1 : struct
+        {
+            internal readonly Timer<T1> Timer;
+
+            private readonly int ticks;
+
+            internal TimeMarker(Timer<T1> timer)
+            {
+                this.Timer = timer;
+                this.ticks = Environment.TickCount;
```

Comment:
This `Environment.TickCount` seems to be a low precision and high cost approach?

Reply:
IDK to be honest.

I will refactor this portion to use StopWatch class. I was just trying to avoid an allocation using Environment.TickCount. But, I just realized the StopWatch class has a static GetTimestamp() and Frequency method.